### DISCUSSION TITLE
Tasmota development, Adding a new sensor...

### DISCUSSION THREAD

**Discussion by trlafleur** [arendst/Tasmota#13242] [Q&A] [answered]
To start with, I'm new to Tasmota development... so I'm sure this is a beginner's question...

I want to develop a new sensor based on "Counters" to measure flow from K-Offset and GPM water meters.
So to get started I copied "Counters", converting it to "FLOW" as my base, without adding any new code. I added it as xsns_125.

It compiles just fine, but when I assign a GPIO pin to FLOW, and I reset the module, I get both FLOW1 and Counter1 displays. (Also, toggling the GPIO pin is not increment the counters... ( works just fine if using just Counters))

So it looks like I missed something very basic...

<img width="393" alt="image" src="https://user-images.githubusercontent.com/4571691/135285155-f9205ba5-f517-4c48-a4e2-e2506a02869a.png">

Made changes in: my_user_config.h to enable it.
~~~
#define USE_Flow
~~~
added globals to settings.h  ( same as counters, but set to FLOW)
~~~
unsigned long   flow_pulse_counter[MAX_COUNTERS];  // 5C0 
uint16_t flow_pulse_counter_type;           // 5D0 
uint16_t flow_pulse_counter_debounce;       // 5D2 
// Only 32 bit boundary variables below
uint16_t flow_pulse_counter_debounce_low;   // FB8 
uint16_t flow_pulse_counter_debounce_high;  // FBA
~~~
added entry in tasmota_template.h
~~~
 D_SENSOR_FLOW "|" D_SENSOR_FLOW "_n|"  // above "Counters"
~~~
added entry in support_feature.ino
~~~
#ifdef USE_FLOW
   feature8 |= 0x00008000; 
#endif
~~~
added entry in en.GB.h
~~~
#define D_SENSOR_FLOW          "Flow Counter"
~~~

Code:
~~~
/*
  xsns_125_flow_ounter.ino - Flow sensors (water meters... sensor support for Tasmota

  Tom@lafleur.us
  Copyright (C) 2021  Maarten Damen and Theo Arends

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#ifdef USE_Flow

//#define MAX_FLOW_COUNTERS 4
#define D_FLOW "Flow"


/*   These need to be global... --> In settings.h
unsigned long   flow_pulse_counter[MAX_COUNTERS];  // 5C0   <---------------  TRL
uint16_t        flow_pulse_counter_type;           // 5D0
uint16_t        flow_pulse_counter_debounce;       // 5D2

// 32bit...
uint16_t        flow_pulse_counter_debounce_low;   // FB8   <----------- TRL
uint16_t        flow_pulse_counter_debounce_high;  // FBA
*/



/*********************************************************************************************\
 * Flow sensors (water meters, unite per minute or K-Offset types...
\*********************************************************************************************/

#define XSNS_125            125

#define D_PRFX_FLOW                 "Flow"
#define D_CMND_FLOW_TYPE            "Type"
#define D_CMND_FLOW_DEBOUNCE        "Debounce"
#define D_CMND_FLOW_DEBOUNCELOW     "DebounceLow"
#define D_CMND_FLOW_DEBOUNCEHIGH    "DebounceHigh"

const char kFlowCommands[] PROGMEM = D_PRFX_FLOW "|"    // Prefix
  "|" D_CMND_FLOW_TYPE "|" D_CMND_FLOW_DEBOUNCE  "|" D_CMND_FLOW_DEBOUNCELOW "|" D_CMND_FLOW_DEBOUNCEHIGH ;

void (* const FlowCommand[])(void) PROGMEM = {
  &CmndFlow, &CmndFlowType, &CmndFlowDebounce, &CmndFlowDebounceLow, &CmndFlowDebounceHigh };

uint8_t flow_ctr_index[MAX_COUNTERS] =  { 0, 1, 2, 3 };

struct FLOW {
  uint32_t  timer[MAX_COUNTERS];            // Last counter time in micro seconds
  uint32_t  timer_low_high[MAX_COUNTERS];   // Last low/high counter time in micro seconds
  uint8_t   no_pullup = 0;                       // Counter input pullup flag (1 = No pullup)
  uint8_t   pin_state = 0;                       // LSB0..3 Last state of counter pin; LSB7==0 IRQ is FALLING, LSB7==1 IRQ is CHANGE
  bool      any_counter = false;
} Flow;

void IRAM_ATTR FlowIsrArg(void *arg) {
  uint32_t index = *static_cast<uint8_t*>(arg);

  uint32_t time = micros();
  uint32_t debounce_time;

  if (Flow.pin_state) 
  {
    // handle low and high debounce times when configured
    if (digitalRead(Pin(GPIO_CNTR1, index)) == bitRead(Flow.pin_state, index)) 
    {
      // new pin state to be ignored because debounce time was not met during last IRQ
      return;
    }
    debounce_time = time - Flow.timer_low_high[index];
    if bitRead(Flow.pin_state, index) {
      // last valid pin state was high, current pin state is low
      if (debounce_time <= Settings->flow_pulse_counter_debounce_high * 1000) return;
    } else {
      // last valid pin state was low, current pin state is high
      if (debounce_time <= Settings->flow_pulse_counter_debounce_low * 1000) return;
    }
    // passed debounce check, save pin state and timing
    Flow.timer_low_high[index] = time;
    Flow.pin_state ^= (1<<index);
    // do not count on rising edge
    if bitRead(Flow.pin_state, index) 
    {
      // PWMfrequency 100
      // restart PWM each second (german 50Hz has to up to 0.01% deviation)
      // restart initiated by setting Counter.startReSync = true;

      return;
    }
  }

  debounce_time = time - Flow.timer[index];
  if (debounce_time > Settings->flow_pulse_counter_debounce * 1000) {
    Flow.timer[index] = time;
    if (bitRead(Settings->flow_pulse_counter_type, index)) {
      RtcSettings.pulse_counter[index] = debounce_time;
    } else {
      RtcSettings.pulse_counter[index]++;
    }
  }
}

/********************************************************************************************/

void FlowInterruptDisable(bool state) {
  if (state) {                                      // Disable interrupts
    if (Flow.any_counter) {
      for (uint32_t i = 0; i < MAX_COUNTERS; i++) {
        if (PinUsed(GPIO_CNTR1, i)) {
          detachInterrupt(Pin(GPIO_CNTR1, i));
        }
      }
      Flow.any_counter = false;
    }
  } else {                                           // Enable interrupts
    if (!Flow.any_counter) {
      FlowInit();
    }
  }
}

bool FlowPinState(void)
{
  if ((XdrvMailbox.index >= AGPIO(GPIO_CNTR1_NP)) && (XdrvMailbox.index < (AGPIO(GPIO_CNTR1_NP) + MAX_COUNTERS))) {
    bitSet(Flow.no_pullup, XdrvMailbox.index - AGPIO(GPIO_CNTR1_NP));
    XdrvMailbox.index -= (AGPIO(GPIO_CNTR1_NP) - AGPIO(GPIO_CNTR1));
    return true;
  }
  return false;
}

void FlowInit(void)
{
  for (uint32_t i = 0; i < MAX_COUNTERS; i++) {
    if (PinUsed(GPIO_CNTR1, i)) {

      Flow.any_counter = true;
      pinMode(Pin(GPIO_CNTR1, i), bitRead(Flow.no_pullup, i) ? INPUT : INPUT_PULLUP);
      if ((0 == Settings->flow_pulse_counter_debounce_low) && (0 == Settings->flow_pulse_counter_debounce_high) ) {
        Flow.pin_state = 0;
        attachInterruptArg(Pin(GPIO_CNTR1, i), FlowIsrArg, &flow_ctr_index[i], FALLING);
      } else {
        Flow.pin_state = 0x8f;
        attachInterruptArg(Pin(GPIO_CNTR1, i), FlowIsrArg, &flow_ctr_index[i], CHANGE);
      }
    }
  }
}

void FlowEverySecond(void)
{
  for (uint32_t i = 0; i < MAX_COUNTERS; i++) {
    if (PinUsed(GPIO_CNTR1, i)) {
      if (bitRead(Settings->flow_pulse_counter_type, i)) {
        uint32_t time = micros() - Flow.timer[i];
        if (time > 4200000000) {  // 70 minutes
          RtcSettings.pulse_counter[i] = 4200000000;  // Set Timer to max in case of no more interrupts due to stall of measured device
        }
      }
    }
  }
}

void FlowSaveState(void)
{
  for (uint32_t i = 0; i < MAX_COUNTERS; i++) {
    if (PinUsed(GPIO_CNTR1, i)) {
      Settings->flow_pulse_counter[i] = RtcSettings.pulse_counter[i];
    }
  }
}

void FlowShow(bool json)
{
  bool header = false;
  uint8_t dsxflg = 0;
  for (uint32_t i = 0; i < MAX_COUNTERS; i++) {
    if (PinUsed(GPIO_CNTR1, i)) {
      char flow[33];
      if (bitRead(Settings->flow_pulse_counter_type, i)) {
        dtostrfd((double)RtcSettings.pulse_counter[i] / 1000000, 6, flow);
      } else {
        dsxflg++;
        snprintf_P(flow, sizeof(flow), PSTR("%lu"), RtcSettings.pulse_counter[i]);
      }

      if (json) {
        if (!header) {
          ResponseAppend_P(PSTR(",\"FLOW\":{"));
        }
        ResponseAppend_P(PSTR("%s\"C%d\":%s"), (header)?",":"", i +1, flow);
        header = true;
#ifdef USE_DOMOTICZ
        if ((0 == TasmotaGlobal.tele_period) && (1 == dsxflg)) {
          DomoticzSensor(DZ_COUNT, RtcSettings.pulse_counter[i]);
          dsxflg++;
        }
#endif  // USE_DOMOTICZ
        if ((0 == TasmotaGlobal.tele_period ) && (Settings->flag3.counter_reset_on_tele)) {
          RtcSettings.pulse_counter[i] = 0;
        }
#ifdef USE_WEBSERVER
      } else {
        WSContentSend_PD(PSTR("{s}" D_FLOW "%d{m}%s%s{e}"),
          i +1, flow, (bitRead(Settings->flow_pulse_counter_type, i)) ? " " D_UNIT_SECOND : "");
#endif  // USE_WEBSERVER
      }
    }
  }
  if (header) {
    ResponseJsonEnd();
  }
}



/*********************************************************************************************\
 * Commands
\*********************************************************************************************/

void CmndFlow(void)
{
  if ((XdrvMailbox.index > 0) && (XdrvMailbox.index <= MAX_COUNTERS)) {
    if ((XdrvMailbox.data_len > 0) && PinUsed(GPIO_CNTR1, XdrvMailbox.index -1)) {
      if ((XdrvMailbox.data[0] == '-') || (XdrvMailbox.data[0] == '+')) {
        RtcSettings.pulse_counter[XdrvMailbox.index -1] += XdrvMailbox.payload;
        Settings->flow_pulse_counter[XdrvMailbox.index -1] += XdrvMailbox.payload;
      } else {
        RtcSettings.pulse_counter[XdrvMailbox.index -1] = XdrvMailbox.payload;
        Settings->flow_pulse_counter[XdrvMailbox.index -1] = XdrvMailbox.payload;
      }
    }
    ResponseCmndIdxNumber(RtcSettings.pulse_counter[XdrvMailbox.index -1]);
  }
}

void CmndFlowType(void)
{
  if ((XdrvMailbox.index > 0) && (XdrvMailbox.index <= MAX_COUNTERS)) {
    if ((XdrvMailbox.payload >= 0) && (XdrvMailbox.payload <= 1) && PinUsed(GPIO_CNTR1, XdrvMailbox.index -1)) {
      bitWrite(Settings->flow_pulse_counter_type, XdrvMailbox.index -1, XdrvMailbox.payload &1);
      RtcSettings.pulse_counter[XdrvMailbox.index -1] = 0;
      Settings->flow_pulse_counter[XdrvMailbox.index -1] = 0;
    }
    ResponseCmndIdxNumber(bitRead(Settings->flow_pulse_counter_type, XdrvMailbox.index -1));
  }
}

void CmndFlowDebounce(void)
{
  if ((XdrvMailbox.payload >= 0) && (XdrvMailbox.payload < 32001)) {
    Settings->pulse_counter_debounce = XdrvMailbox.payload;
  }
  ResponseCmndNumber(Settings->flow_pulse_counter_debounce);
}

void CmndFlowDebounceLow(void)
{
  if ((XdrvMailbox.payload >= 0) && (XdrvMailbox.payload < 32001)) {
    Settings->flow_pulse_counter_debounce_low = XdrvMailbox.payload;
    FlowInit();
  }
  ResponseCmndNumber(Settings->flow_pulse_counter_debounce_low);
}

void CmndFlowDebounceHigh(void)
{
  if ((XdrvMailbox.payload >= 0) && (XdrvMailbox.payload < 32001)) {
    Settings->flow_pulse_counter_debounce_high = XdrvMailbox.payload;
    FlowInit();
  }
  ResponseCmndNumber(Settings->flow_pulse_counter_debounce_high);
}

/*********************************************************************************************\
 * Interface
\*********************************************************************************************/

bool Xsns125(uint8_t function)
{
  bool result = false;

  if (Flow.any_counter) {
    switch (function) {
      case FUNC_EVERY_SECOND:
        FlowEverySecond();
        break;
      case FUNC_JSON_APPEND:
        FlowShow(1);
        break;

#ifdef USE_WEBSERVER
      case FUNC_WEB_SENSOR:
        FlowShow(0);
        break;
#endif  // USE_WEBSERVER

      case FUNC_SAVE_BEFORE_RESTART:
      case FUNC_SAVE_AT_MIDNIGHT:
        FlowSaveState();
        break;
      case FUNC_COMMAND:
        result = DecodeCommand(kFlowCommands, FlowCommand);
        break;
    }
  } else {
    switch (function) {
      case FUNC_INIT:
        FlowInit();
        break;
      case FUNC_PIN_STATE:
        result = FlowPinState();
        break;
    }
  }
  return result;
}

#endif  // USE_FLOW


~~~

**Comment by barbudor** [comment, marked as answer]
Hi
It could be easier if you just provide the link to your fork/branch so we can see exactly what you changes
Touching the Settings structure is not recommended for a newcomer in tasmota development. I would recommend starting your dev and debug without it.

If you are using GPIO_CNT don't be surprised to have Counter displayed (or disable it if you intend FLOW to be exclusive of COUNTER)
You should define a new type of GPIO such as GPIO_FLOW

I'll look at the rest later

**Reply by barbudor** [reply]
You seem to have broken a few rules

- You should have a new GPIO type such as GPIO_FLOW1 and GPIO_FLOW1_NP
- New GPIO_XXX should be added 
  - at the end of `enum UserSelectablePins { }` just before GPIO_SENSOR_END 
  - at the end of `const char kSensorNames[]`, that you have done
  - in `const uint16_t kGpioNiceList[]` at a convenient place, in your case close to GPIO_CNTR1

I'm not sure what you are trying to achieve but if the behavior is close to a counter, why not looking at improving the existing counter by adding a new counter type ? (although there is some challenge to add additional type without introducing a breaking change in Settings but that is feasible in 2 steps)

**Comment by trlafleur** [comment]
Thanks... will make these changes, will post the full project on GitHub later today...
Beginners errors!!

**Comment by trlafleur** [comment]
Ok, I did make the changes, but there is still some issue, when I select GPIO to FLOW, and save config, it restarts with Counter1... also, not counting if I toggle the pin...  so something else is wrong in my basic setup...

https://github.com/trlafleur/Tasmota/tree/tasmota-local

thanks

**Comment by trlafleur** [comment]
Looks like I may need to add a FLOW_MAX_COUNTERS

settings.ino    
~~~
bool read_valid = (RTC_MEM_VALID == RtcSettings.valid);
  if (update) {
    if (!read_valid) {
      memset(&RtcSettings, 0, sizeof(RtcSettings));
      RtcSettings.valid = RTC_MEM_VALID;
      RtcSettings.energy_kWhtoday = Settings->energy_kWhtoday;
      RtcSettings.energy_kWhtotal = Settings->energy_kWhtotal;
      RtcSettings.energy_usage = Settings->energy_usage;
      for (uint32_t i = 0; i < MAX_COUNTERS; i++) {
        RtcSettings.pulse_counter[i] = Settings->pulse_counter[i];
      }
~~~

**Reply by barbudor** [reply]
Having a dedicated FLOW_MAX_COUNTER may help undesrtanding the code but it it's not intended to be different than MAX_COUNTER it's not a problem

I will pull your code later today, that will be easier to have a look and may be test it too

**Reply by barbudor** [reply]
Quite some broken things in GPIO and Settings
I will PR some fixes to you soon

**Reply by barbudor** [reply]
Please see https://github.com/trlafleur/Tasmota/pull/1
I made some changes to cleanup Settings, Templates.....

**Reply by barbudor** [reply]
Regarding your question about debug messages, the standard Tasmota way is `AddLog()`
```
AddLog( LOG_LEVEL_xxx, PSTR("XXX: some sentence %d %f"), var1, var2);
```
Logs can be reported on 4 different channels : Web console, Serial console, Syslog and MqttLog
There are 4 levels: LOG_LEVEL_ERROR (1), LOG_LEVEL_INFO (2), LOG_LEVEL_DEBUG (3), LOG_LEVEL_DEBUG_MORE (3) matching the number in the commands `SerialLog <n>` or `WebLog <n>` or `Syslog <n>` or `MqttLog<n>`
Logs on a given channel will be printed/emitted if the LOG_LEVEL_xxx specified in `AddLog()` is lower or equal than value specified in the command (default to 2 for Serial and Web, and 0 == none for Syslog and MQTT)
Don't forget that logs are asynchronous. AddLog() place the log message in buffer which is then sent in the background. It has advantage in terms of real-time (the call to the function doesn't take much time) but in you have a crash before logs are emitted, it makes it difficult to find where the problem occurred.
I use a lot serial logs when debugging on NodeMCU/D1-mini boards. When that's not available, console may be ok, or syslog (I'm using a simple pseudo syslog server in 5 lines of python that prints on screen what it receives).
When sending a PR to Tasmota, comment all logs are are meant to debug the code. Only keep active logs that can be useful for a user to debug its usage of the feature.
_"Every byte counts"_

**Comment by trlafleur** [comment]
Ok, I'm a bit further along in this project and have some questions...
The code is very early at this stage and is more of an experiment in learning Tasmota!
~~~
https://github.com/trlafleur/Tasmota/tree/tasmota-local
~~~

1) is there a way to tell if it's the first boot after a compile?
2) Is there information on the macro {s}, {m}... etc?? I think I know what they do, but??
3) how do you load a template at compile time... my tries did not work by adding this to user_config_override.h
~~~
#undef USER_TEMPLATE
  #define USER_TEMPLATE "{\"Flow-Test\":\"ESP32-DevKit\",\"GPIO\":[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,0,1,1,1,0,0,0,0,1,1,1,1,1,0,0,1],\"FLAG\":0,\"BASE\":1}"  // [Template] Set JSON template
 ~~~

4) Not sure yet on how to properly format an MQTT JSON message...
~~~
char flow_period[33];     // neeed to adjust size and number of these... <-----------  TRL
      char flow_count[33];
      char flow_1hrVolume[33];
      char flow_24hrVolume[33];
      char flow_RateFactor[33];
      char flow_k[33];
      char flow_offset[33];
      char flow_rate[33];

        dtostrfd((double)RtcSettings.flow_pulse_counter_period[i] / 1000000, 6, flow_period);
        dtostrfd((double)Flow.rate_factor[i], 3, flow_RateFactor);
        dtostrfd((double)Flow.K[i], 3, flow_k);
        dtostrfd((double)Flow.Offset[i], 3, flow_offset);
        dtostrfd((double)CurrentFlowRate[i] * rate_factor, 3, flow_rate);
  
        dsxflg++;
        snprintf_P(flow_count, sizeof(flow_count), PSTR("%lu"), RtcSettings.flow_pulse_counter[i]);
        snprintf_P(flow_1hrVolume, sizeof(flow_1hrVolume), PSTR("%lu"), Saved1hrPulseVolume[i]);
        snprintf_P(flow_24hrVolume, sizeof(flow_24hrVolume), PSTR("%lu"), Saved24hrPulseVolume[i]);
        //snprintf_P(flow_k, sizeof(flow_k), PSTR("%lu"), Flow.rate_factor[i]);
        //snprintf_P(flow_offset, sizeof(flow_offset), PSTR("%lu"), RtcSettings.flow_pulse_counter[i]);

      if (json) 
      {
        if (!header) 
        {
          ResponseAppend_P(PSTR(",\"FLOW\":{"));
        }
        ResponseAppend_P(PSTR("%s\"FlowCount%d\": %s, "),   (header)?",":"", i +1, flow_count);
        ResponseAppend_P(PSTR("%s\"Flow%d\": %s, "),        (header)?",":"", i +1, flow_rate);
        ResponseAppend_P(PSTR("%s\"FlowPeriod%d\": %s, "),  (header)?",":"", i +1, flow_period);
        ResponseAppend_P(PSTR("%s\"1HrVolume%d\": %s, "),   (header)?",":"", i +1, flow_1hrVolume);
        ResponseAppend_P(PSTR("%s\"24HrVolume%d\": %s, "),  (header)?",":"", i +1, flow_24hrVolume);
        ResponseAppend_P(PSTR("%s\"Rate Factor%d\": %s, "), (header)?",":"", i +1, flow_RateFactor);
        ResponseAppend_P(PSTR("%s\"K%d\":%s"),              (header)?",":"", i +1, flow_k);
        ResponseAppend_P(PSTR("%s\"FlowUnits%d\":%s"),      (header)?",":"", i +1, Current_Units);


        //ResponseAppend_P(PSTR("%s\"FlowCount%d\":%s"), (header)?",":"", i +1, flow_count);
        header = true;

~~~

5) Just to be clear, if I need to add some variables in settings.h, I need to be aware of the boundaries of data structures,
and care in selection locations for them...
~~~
 volatile uint32_t flow_pulse_counter_period[MAX_COUNTERS];     // <-------------------  TRL
  volatile uint32_t flow_pulse_counter[MAX_COUNTERS];

                                           // 2F0 - 2FF free locations // note This <-------------------  TRL


uint32_t      flow_pulse_counter[MAX_COUNTERS];  // 560 4x4=16  <---------------  TRL
uint16_t      flow_pulse_counter_type = 0;           // 570
uint16_t      flow_pulse_counter_debounce_low = 0;   // 574   <----------- TRL
uint16_t      flow_pulse_counter_debounce_high = 0;  // 576
uint16_t      flow_pulse_counter_debounce = 0;      // 578

  uint8_t       free_56C[66];              // 578   // <---- from 68 to 66  TRL
~~~

6) What variables are saved and re-loaded when I do a Backup-Restore configuration?

7) What is the preferred way to set default values at start-up...

**Reply by barbudor** [reply]
So many questions 🤣

1) is there a way to tell if it's the first boot after a compile?

No. Generally what's counting is the 1st boot after a flashing from an erased flash.
In that case, tasmota will initialize settings with default values that you may have overriden in `user_config_override.h`
If you upgrade a device already configured, it keeps the existing configuration.
You can issue a `reset 1` to erase current settings and force a reload of all default settings like a 1st boot after flash erase

2) Is there information on the macro {s}, {m}... etc?? I think I know what they do, but??

These are placeholders for standard formating (colours, position etc...) that will be replaced by parameters as configured by WebColors for exemple. You must respect the same format for your Web insertion in order to keep the look and feel of the GUI

3) how do you load a template at compile time... my tries did not work by adding this to user_config_override.h

2 things:
- `#define USER_TEMPLATE ...` as you did 
- `#define MODULE USER_MODULE` verbatim as this. It tells tasmota that the active module is the USER_MODULE from the the USER_TEMPLATE

As said in 1, this will be enforced only on a 1st flash or following a `reset 1`


That's the first set of answers. I will come back soon with the others

**Reply by barbudor** [reply]
4) Not sure yet on how to properly format an MQTT JSON message...

I don't see the whole code here. But I assume there is a loop for all Flow indexed by `i`.
So the header `, "Flow": { ` should be  present only once or can be inserted before the loop.
May be I should have a look a the whole code if your repository is up to date

5) if I need to add some variables in settings.h

- Put your settings in a single place, not spread all over the structure.
- You must remove as much bytes from the `free_xxx[nn]` as your variables are using
- There are alignment consideration and the compiler may force some alignement you must take into account as a variable will ALWAYS be assigned to an offset multiple of it's size (an float or (u)int32_t will always start on a boundary multiple of 4, an (u)int16_t will always start on a even boundary. So the best rule to minimize bytes lost by alignment is to have all 4-bytes-sized variables first, then the 2-bytes-sized, then the 1-byte-sized.
- The comments should provide the offset on each line

I can't understand the 3 first lines:
- the 2 first lines have no offset in comments
- the 3rd line have comment where the offset values are not in continuity of the following lines
- I don't understand "<---- from 68 to 66  TRL"

6) What variables are saved and re-loaded when I do a Backup-Restore configuration?

The whole `Settings` structure

7) Most of the code for that is in `settings.ino` 

It's a quite some complex piece of code because it has to deal with multiple cases including
a) Starting on a freshly flashed device where there is no previoous settings
b) Upgrading a device previously configured but those variables were not existant so they start with default value 0
c) Upgrading from a version with significant changes in the Settings structure that requires relocation of settings

a) is the simplest
b) your code must support to start with variables set as 0 and may be identify this properly

**Reply by barbudor** [reply]
4) Not sure yet on how to properly format an MQTT JSON message...

Looking at `[void FlowShow(bool json)](https://github.com/trlafleur/Tasmota/blob/78aea747ee69ac30e52e0ad151e7fb6505336f6d/tasmota/xsns_125_flow_counter.ino#L435)`

Don't use `double`. We don't want `double` math in Tasmota because it takes up too much resources. Stick with `float`
https://github.com/trlafleur/Tasmota/blob/78aea747ee69ac30e52e0ad151e7fb6505336f6d/tasmota/xsns_125_flow_counter.ino#L453-L457

Your logic is wrong
https://github.com/trlafleur/Tasmota/blob/78aea747ee69ac30e52e0ad151e7fb6505336f6d/tasmota/xsns_125_flow_counter.ino#L468-L483

You should insert the header `, "Flow":{` only once as far as there is at least 1 "Flow" configured

You should always add a `,` separator between the variables for a give flow
But only add one if there is another flow.
Always insert a `,` between `FlowCount` and each item until `FlowUnits`

My suggestion would be (pseudo-code, not true C)
```
bool header = false
for i = 0 to effective number of flows -1 (so you only issue data for  flows is 2 are configured)
  if (!header) 
    insert `,"Flow":{`    // there is always a `,` here before `"Flow"` because there is always at least a "Time" element
  else 
    insert `,`
  insert FlowCount<i> with an `,` at the end
  insert Flow<i> with an `,` at the end
  ...
  insert FlowUnits<i> BUT WITHOUT `,` at the end (it will be insert at the beginning of next loop
  header = true;  make next loop insert a `,` separator instead of `"Flow":{`
end for
if header (which means we've een through the loop at least once)
  insert `}` to close the `"Flow":{`
```

Remove ALL unecessary spaces in your JSON. There are useless for JSON parsing.
Every byte count, in Flash and in RAM buffers

**Reply by barbudor** [reply]
5. if I need to add some variables in settings.h

Now I understand, the top 3 lines refers to RTC and the rest to Settings
```
uint32_t      energy_kWhtotal_ph[3];     // 2E4

  volatile uint32_t flow_pulse_counter_period[MAX_COUNTERS];     // <-------------------  TRL
  volatile uint32_t flow_pulse_counter[MAX_COUNTERS];

                                           // 2F0 - 2FF free locations // note This <-------------------  TRL
} TRtcSettings;
```
Totally wrong
You are adding 8 uint32_t for a total of 32 bytes fromo 0x2E4 to an area that ends at 0x2FF, so only 28 bytes free
You need to review your expectations


As I said from the 1st time, your code is going to be very costly as it almost duplicate the Counter feature
You shoud consider implementing as either an alternative (exclusive to the usage of Counter) or as an extension (like a new CounterType)

**Comment by trlafleur** [comment]
my github is up to date

~~ _/) ~~~~ _/) ~~~~ _/) ~~~~ _/) ~~

Tom Lafleur

**Comment by trlafleur** [comment]
I have question...

I want to send an async message when a variable is set, without out waiting
for the next message queue time??
is there a recommended way??
is there a call back to let me know it was sent??

thanks


~~ _/) ~~~~ _/) ~~~~ _/) ~~~~ _/) ~~

Tom Lafleur

**Reply by barbudor** [reply]
Tom,
Sorry I don't get what you aer looking for
Could you please elaborate ?

**Comment by trlafleur** [comment]
Tasmota sends a sensor message at a rate determined by some setting....
I from time to time, want to send a MQTT message when a values changes and
not wait for the next sensorTX window
If the process is async, it would be nice to have a call-back or flag to
tell me message has been sent...

Is this possible?

~~ _/) ~~~~ _/) ~~~~ _/) ~~~~ _/) ~~

Tom Lafleur

**Comment by barbudor** [comment]
Sensor message is sent at Teleperiod. Basic Tasmota feature

But driver's read function is performed more often
You can use rules to determine that a value has changed and either send a custom message with  the command publish
https://tasmota.github.io/docs/Rules/#transmit-sensor-value-only-when-a-delta-is-reached
